### PR TITLE
rtai: don't unload rtai kernel modules

### PR DESCRIPTION
**Pull request by jepler**
A rather long time ago now, we found that there was a low probability chance of a full system crash when unloading linuxcnc on rtai systems, as reported here: https://mail.rtai.org/pipermail/rtai/2020-June/028122.html -- any record of the specific oops seems to have gone missing, but I believed it was specific to unloading the rtai modules.

From what I recall, the chance of lockup seemed to be higher on a vm than on bare metal, but I definitely produced the minimized test case that Andy posted and verified it on real, non-emulated hardware.

This is one of the factors that led to us de-emphasizing rtai based builds, because we couldn't reliably test them on buildbot.

Anyway, I was musing on the fact that at least uspace doesn't actually require these modules to be unloaded between runs, as far as I knew. So anyway, this patch removes the module unloading steps from the realtime script for both uspace+rtapi and kmod+rtapi.

uspace+rtai and kmod+rtai both ran 'halrun test.hal' of tests/abs.0 over 30k times each with no full-system crash.

The original reproducer (unloading rtai_sched) crashed at ~500 cycles in the VM.

This test setup details: host is debian bullseye on ryzen 5700u using kvm virtualization, guest is debian bullseye with 4 cores, 4GB RAM, kernel 4.19.195-RTAI

About "uspace+rtai", which is not the mode that most people use linuxcnc and rtai in: Personally, for a number of reasons I'd like to end up promoting this mode over the old "kmod+rtai" mode if/as we bring back rtai kernel support, but that's contingent on the realtime performance on hardware actually being at par with the older system.

(Short version of the argument: Crashing within kernel modules can easily take down the whole system; uspace+rtai is more forgiving.  No need to make special math libraries for kernel space. the possibility of making "uspace+rtai" a very small deb that augments the regular uspace deb, rather than two separate builds)

### CODE REVIEW

**Comment by jepler** [comment]
After 8654 iterations, it failed with "rtapi_shmem_new failed due to shmget(key=0x48414c32): Permission denied" but the system remained usable. There appears to be a race condition in `rtapi_shmem_new`'s uspace implementation; the segment is created as root if rtapi_app creates it with `shmget(..., IPC_CREAT|...)`, and then the owner is changed with `shmctl(..., IPC_SET)`. In between the segment briefly exists but cannot be attached by another program such has halcmd. It's unclear why that function is even run WITH_ROOT...

**Comment by jepler** [comment]
This later passed 35,000 iterations without problem, at which point I ended the test.

Further notes on the tested system: Host is a Ryzen 5700U running Debian Bullseye.  Guest is KVM hypervisor, 4GB RAM, 4 CPU cores allocated to the VM.  Also running bullseye, but with the 4.19.195-RTAI kernel from linuxcnc.org installed. Locally built LinuxCNC, running says "Note: Using LXRT realtime", so it's using the realtime code paths I expect.

**Comment by jepler** [comment]
It's also important to verify that the original test case actually crashes the system too.. I ran the reproducer from https://mail.rtai.org/pipermail/rtai/2020-June/028122.html and after ~500 iterations it crashed the whole system with no useful messages on the console.
image
this gives me good confidence that the strategy of never unloading modules will work around the apparent rtai bug.

**Comment by jepler** [comment]
A bit later this truncated kernel backtrace was displayed
image

**Comment by jepler** [comment]
I let kmod+rtai pass 536,000 iterations (half a million) and it's still going strong.

File: scripts/realtime.in [hunk 1]
```diff
@@ -103,10 +103,13 @@ CheckConfig(){
         MODULES_UNLOAD="$MOD $MODULES_UNLOAD"
     done
     case $RTPREFIX in
-    *rtai*)
+    (*rtai*)
```
Comment (petterreinholdtsen) [inline scripts/realtime.in:106]: This use of '(' is quite unusual in a case/esac block.  Why do you want it?

**Comment by petterreinholdtsen** [comment]
[Jeff Epler]
> There appears to be a race condition in `rtapi_shmem_new`'s uspace
> implementation; the segment is created as root if rtapi_app creates it
> with `shmget(..., IPC_CREAT|...)`, and then the owner is changed with
> `shmctl(..., IPC_SET)`. In between the segment briefly exists but
> cannot be attached by another program such has halcmd. It's unclear
> why that function is even run WITH_ROOT...

If the problem only exist for a brief moment, why not make halcmd and
others try again if the connection fail, and only give up after trying a
few iterations?

-- 
Happy hacking
Petter Reinholdtsen

**Comment by andypugh** [comment]
This looks very promising, and thinking about it, it doesn't really seem necessary or even advisable to unload RTAI modules when LinuxCNC exits. Most machines will start up, load LinuxCNC, then shut down anyway, but even in use-cases where LinuxCNC is loaded and unloaded several times in a session there seems to be little penalty in leaving the modules loaded?

File: scripts/realtime.in (hunk 1 above)
Comment (jepler) [reply scripts/realtime.in:106]: Naive text editors (those that perform paren matching but don't fully understand language syntax) always want parens to match. It's just a stylistic thing and can be dropped here if it's controversial.

**Comment by petterreinholdtsen** [comment]
[Jeff Epler]
> Naive text editors (those that perform paren matching but don't fully
> understand language syntax) always want parens to match. It's just a
> stylistic thing and can be dropped here if it's controversial.

I do not really believe it is controversial.  It is simply not according
to the bourne shell syntax, aka wrong.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by jepler** [comment]
The optional beginning "(" is part of the shell command language as specified by opengroup: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_04_05 and is suggested as a technique when writing shell `case` statements in configure.ac scripts so while this is not a .ac / .m4 script, its real world portability is probably not in doubt https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/Balancing-Parentheses.html and anyway the script is `#!/usr/bin/bash` so bashisms are OK even if Solaris 10 (released 2005) doesn't support it in /bin/sh.

I'm sorry this detail is crowding out any discussion of the actual useful change :-/

**Comment by petterreinholdtsen** [comment]
I stand corrected.  If your aim is bash and not bourne shell, () can be
used.  I try to stay with bourne shell notation as /bin/dash on Debian
is /bin/sh, and dash is a lot quicker to start than /bin/bash.

> I'm sorry this detail is crowding out any discussion of the actual
> useful change :-/

I do not know the real time stuff in Linux enough to have an opinion
here.  From what I read so far in this thread, I suspect there is still
a race condition in starting up linuxcnc, which is not removed by
dropping the unload calls.  I suspect that race is fixable by adding
some retries when trying to connect to the shared memory.

-- 
Happy hacking
Petter Reinholdtsen

**Comment by jepler** [comment]
dash accepts the syntax
```
$ dash -c 'case "$0" in (y*) echo aye ;; (*) echo nay ; esac' yes
aye
$ dash -c 'case "$0" in (y*) echo aye ;; (*) echo nay ; esac' no
nay
```

**Comment by jepler** [comment]
where this would be irritating is switching to an older linuxcnc package -- at exit, halrun left rtai_ modules loaded, but old linuxcnc will treat this as "realtime still running".  It's unfortuante, but I don't know what the workaround or fix would be, since it's all about the behavior of an older version.

**Review by SebKuzminsky — APPROVED** [review]
Thanks for these fixes @jepler !

**Comment by andypugh** [comment]
I will try pulling master and running the torture test on real hardware. But not for a few days.

**Pull request merged**